### ISSUE TITLE
starting container process caused: "exec: \"/bin/sh\": stat /bin/sh: no such file or directory"

### ISSUE THREAD

**Issue by Albert221** [moby/moby#31702]
**Description**
I have error after running `RUN apt -yqq install nginx iputils-ping`:

> container_linux.go:247: starting container process caused: "exec: \\"/bin/sh\\": stat /bin/sh: no such file or directory"
> ERROR: Service 'nginx' failed to build: oci runtime error: \<same as above>

But the weird thing is, that when I had Docker installed previously, it never happend (I had Docker, uninstalled it and after some time installed again).

**Steps to reproduce the issue:**
1. Run `docker-compose build`
2. See error

**`Dockerfile`**

```
FROM ubuntu:16.04

WORKDIR /usr/local/src

RUN apt -yqq update
RUN apt -yqq install nginx iputils-ping

COPY docker/nginx/dev.conf /etc/nginx/conf.d/dev.conf
COPY docker/nginx/nginx.conf /etc/nginx/nginx.conf

CMD ["nginx"]
```

**`docker-compose.yml`**

```
version: "3"
services:
  mariadb:
    image: mariadb
    entrypoint: docker-entrypoint.sh mysqld --character-set-server=utf8mb4 --collation-server=utf8mb4_unicode_ci --innodb_large_prefix=1 --innodb_file_format=barracuda --innodb_file_format_max=barracuda --innodb_file_per_table=1
    environment:
      - MYSQL_ALLOW_EMPTY_PASSWORD=1
      - MYSQL_ROOT_PASSWORD=
  phpmyadmin:
    image: phpmyadmin/phpmyadmin
    ports:
      - "8080:80"
    restart: always
    environment:
      - PMA_HOST=mariadb
    links:
      - mariadb
  php:
    image: php:7.1.1-fpm
    ports:
      - "9000:9000"
    volumes:
      - .:/dogopic
    links:
      - mariadb
  nginx:
    build: .
    ports:
      - "8000:80"
    volumes:
      - .:/dogopic
    links:
      - php
```

**Describe the results you received:**
```
PS E:\dogopic> docker-compose build
mariadb uses an image, skipping
phpmyadmin uses an image, skipping
php uses an image, skipping
Building nginx
Step 1/8 : FROM ubuntu:16.04
 ---> 0ef2e08ed3fa
Step 2/8 : MAINTAINER Olgierd Grzyb <kontakt@olgierd.me>
 ---> Using cache
 ---> d00af82c32f0
Step 3/8 : WORKDIR /usr/local/src
 ---> Using cache
 ---> 0a0135a8020e
Step 4/8 : RUN apt -yqq update
 ---> Running in d5a972b3af0d
container_linux.go:247: starting container process caused "exec: \"/bin/sh\": stat /bin/sh: no such file or directory"
ERROR: Service 'nginx' failed to build: oci runtime error: container_linux.go:247: starting container process caused "exec: \"/bin/sh\": stat /bin/sh: no such file or directory"
```

**Describe the results you expected:**
Build containers successfuly.

**Output of `docker version`:**

```
Client:
 Version:      17.03.0-ce
 API version:  1.26
 Go version:   go1.7.5
 Git commit:   60ccb22
 Built:        Thu Feb 23 10:40:59 2017
 OS/Arch:      windows/amd64

Server:
 Version:      17.03.0-ce
 API version:  1.26 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   3a232c8
 Built:        Tue Feb 28 07:52:04 2017
 OS/Arch:      linux/amd64
 Experimental: true
```

**Output of `docker info`:**

```
Containers: 6
 Running: 0
 Paused: 0
 Stopped: 6
Images: 7
Server Version: 17.03.0-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 977c511eda0925a723debdc94d09459af49d082a
runc version: a01dafd48bc1c7cc12bdb01206f9fea7dd6feb70
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.12-moby
Operating System: Alpine Linux v3.5
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 1.934 GiB
Name: moby
ID: XPR4:7B3C:XUDP:VCTK:OOGW:NQJW:A7L3:BSJO:GN6R:KLJU:FKNX:S5V4
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 13
 Goroutines: 28
 System Time: 2017-03-09T18:31:09.1487623Z
 EventsListeners: 0
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Comment by james4388** [comment]
You load the image when docker starting up. Just remove and install docker again. Work for me

**Comment by Albert221** [comment]
I did it multiple times.

**Comment by rawkode** [comment]
Tried to reproduce this and couldn't. Our `docker version` is slightly different though:

Yours:
```
Server:
 Version:      17.03.0-ce
 API version:  1.26 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   3a232c8
 Built:        Tue Feb 28 07:52:04 2017
 OS/Arch:      linux/amd64
 Experimental: true
```

Mine:
```
Server:
 Version:      17.03.0-ce
 API version:  1.26 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   60ccb22
 Built:        Thu Feb 23 11:07:40 2017
 OS/Arch:      linux/amd64
 Experimental: true
```

**Comment by krzkaczor** [comment]
I just encountered a similar problem on mac. https://github.com/docker/for-mac/issues/1396 helped. I also did full docker cleanup first (removed unused images etc.)

**Comment by persus** [comment]
Same problem here with Git commit 3a232c8.

I built my own image by writing an dockerfile.
`FROM localhost:5000/persus/alpine:3.5
MAINTAINER Perica Sušić <persus@susic.info>

# Einvorinment Information
ENV MIRROR_TESTING http://mirror1.hs-esslingen.de/pub/Mirrors/alpine/edge/testing
ENV EMBY_BASE_URL https://github.com/MediaBrowser/Emby/releases/download/
ENV EMBY_VERSION 3.2.10
ENV EMBY_PACKAGE_NAME Emby.Mono
ENV EMBY_PACKAGE ${EMBY_VERSION}/${EMBY_PACKAGE_NAME}.zip
ENV EMBY_PACKAGE_URL ${EMBY_BASE_URL}/${EMBY_PACKAGE}
ENV APP_PATH /opt/emby
ENV APP_CFG_PATH /config
ENV APP_MEDIA_PATH /media

# Install apk packages
RUN echo "@testing ${MIRROR_TESTING}" >> /etc/apk/repositories \
 && apk --update upgrade           \
 && apk --no-cache add             \
        ca-certificates            \
        mono@testing               \
        imagemagick                \
        ffmpeg                     \
        sqlite                     \
        unzip                      \
        wget                       \
 && cert-sync /etc/ssl/certs/ca-certificates.crt       \
# Download and install Emby
 && mkdir -p ${APP_PATH}                               \
 && wget -O "${APP_PATH}/emby.zip" ${EMBY_PACKAGE_URL} \
 && unzip "${APP_PATH}/emby.zip" -d ${APP_PATH}        \
 && rm "${APP_PATH}/emby.zip"                          \
# User and group
 && addgroup -g 1001 -S emby                           \
 && adduser -u 1000 -SHG emby emby                     \
# Config directory
 && mkdir -p ${APP_CFG_PATH}                           \
 && chown -R emby:emby ${APP_CFG_PATH}                 \
# Media directory
 && mkdir -p ${APP_MEDIA_PATH}                         \
 && chown -R emby:emby ${APP_MEDIA_PATH}               \
# Uninstall unneeded packages
 && apk del wget unzip make gcc g++                    \
# Cleanup
 && rm -rf /var/cache/apk/*                            \
 && rm -rf /tmp/*                                      \
 && rm -rf /var/log/*

WORKDIR ${APP_PATH}
VOLUME ["${APP_CFG_PATH}", "${APP_MEDIA_PATH}"]
COPY *.config ${APP_PATH}/
COPY emby.sh /etc/service/emby/run
RUN chmod +x /etc/service/*/run

# Default Emby HTTP/tcp server port
EXPOSE 8096/tcp
# Default Emby HTTPS/tcp server port
EXPOSE 8920/tcp
# UDP server port
EXPOSE 7359/udp
# SSDP port for UPnP / DLNA discovery
EXPOSE 1900/udp

CMD ["/sbin/start_runit"]`

Then I ran the command 
`sudo docker build -t persus/alpine-emby:3.2 -f emby.alpine.build.dockerfile .`

When trying to run the newly built image by typing
`sudo docker run -d --name emby -p 1900:1900/udp -p 8096:8096/tcp -p 7359:7359/udp -p 8920:8920/tcp  persus/alpine-emby:3.2`

I get a similar error message
`docker: Error response from daemon: oci runtime error: container_linux.go:247: starting container process caused "exec: \"/sbin/start_runit\": stat /sbin/start_runit: no such file or directory".`

I am new to docker. Am I doing something wrong?

**Comment by persus** [comment]
Oh I found the error. I set the wrong CMD-command. Should have been "/etc/service/emby/run/start.sh" but was "/sbin/start_runit". After I set the right CMD-command it all worked well. I guess this was a copy'n'paste error since I do experiment with su-exec :(

After I set the right CMD-command I had a bug in the start-script pointing to a file that does not exists. Then I got a similar error (is slightly different).
```
docker: Error response from daemon: oci runtime error: container_linux.go:247: starting container process caused "exec: \"/etc/service/emby/run/emby.sh\": stat /etc/service/emby/run/emby.sh: not a directory"
: Are you trying to mount a directory onto a file (or vice-versa)? Check if the specified host path exists and is the expected type.
docker: Error response from daemon: oci runtime error: container_linux.go:247: starting container process caused "exec: \"/etc/service/emby/run/emby.sh\": stat /etc/service/emby/run/emby.sh: not a directory"
: Are you trying to mount a directory onto a file (or vice-versa)? Check if the specified host path exists and is the expected type.
``` 

So what I learned is: When a given file does not exist you get the "no such file or directory." error message. When a file does not exist in the script, one gets a slightly different message

My dockerfile looks now like this
```
FROM localhost:5000/persus/alpine:3.5
MAINTAINER Perica Sušić <persus@susic.info>

# Einvorinment Information
ENV MIRROR_TESTING http://mirror1.hs-esslingen.de/pub/Mirrors/alpine/edge/testing
ENV EMBY_BASE_URL https://github.com/MediaBrowser/Emby/releases/download/
ENV EMBY_VERSION 3.2.10
ENV EMBY_PACKAGE_NAME Emby.Mono
ENV EMBY_PACKAGE ${EMBY_VERSION}/${EMBY_PACKAGE_NAME}.zip
ENV EMBY_PACKAGE_URL ${EMBY_BASE_URL}/${EMBY_PACKAGE}
ENV APP_PATH /opt/emby
ENV APP_CFG_PATH /config
ENV APP_MEDIA_PATH /media

COPY emby.mono.start.sh /
RUN chmod +x /emby.mono.start.sh   \
# Install apk packages
 && echo "@testing ${MIRROR_TESTING}" >> /etc/apk/repositories \
 && apk --update upgrade           \
 && apk --no-cache add             \
        ca-certificates            \
        mono@testing               \
        imagemagick                \
        ffmpeg                     \
        sqlite                     \
        unzip                      \
        wget                       \
 && cert-sync /etc/ssl/certs/ca-certificates.crt       \
# Download and install Emby
 && mkdir -p ${APP_PATH}                               \
 && wget -O "${APP_PATH}/emby.zip" ${EMBY_PACKAGE_URL} \
 && unzip "${APP_PATH}/emby.zip" -d ${APP_PATH}        \
 && rm "${APP_PATH}/emby.zip"                          \
# User and group
 && addgroup -g 1001 -S emby                           \
 && adduser -u 1000 -SHG emby emby                     \
# Config directory
 && mkdir -p ${APP_CFG_PATH}                           \
 && chown -R emby:emby ${APP_CFG_PATH}                 \
# Media directory
 && mkdir -p ${APP_MEDIA_PATH}                         \
 && chown -R emby:emby ${APP_MEDIA_PATH}               \
# Uninstall unneeded packages
 && apk del wget unzip make gcc g++                    \
# Cleanup
 && rm -rf /var/cache/apk/*                            \
 && rm -rf /tmp/*                                      \
 && rm -rf /var/log/*

WORKDIR ${APP_PATH}
VOLUME ["${APP_CFG_PATH}", "${APP_MEDIA_PATH}"]
COPY *.config ${APP_PATH}/

# Default Emby HTTP/tcp server port
EXPOSE 8096/tcp
# Default Emby HTTPS/tcp server port
EXPOSE 8920/tcp
# UDP server port
EXPOSE 7359/udp
# SSDP port for UPnP / DLNA discovery
EXPOSE 1900/udp

CMD ["/emby.mono.start.sh"]
```

And you sure want the content of the start script "emby.mono.start.sh"
```
#!/bin/sh

/usr/bin/mono /opt/emby/MediaBrowser.Server.Mono.exe \
  -programdata /config    \
  -ffmpeg /usr/bin/ffmpeg \
  -ffprobe /usr/bin/ffprobe
```

**Comment by qubusp** [comment]
same error 
Image looks like that - 

```
# Set nginx base image
FROM gcr.io/mdg-public/engine:1.0.6
COPY testing.json .
RUN ENGINE_CONFIG=$(cat ./testing.json)
```

**Comment by qubusp** [comment]
change the name of the CMD script in my opinion.

**Comment by thaJeztah** [comment]
Looking at the above, I don't think there's a bug at hand, so I'll close this issue.

There are various reasons why this (or similar errors) can occur; some examples:

Create `script.sh`

```bash
cat <<-'EOF'> script.sh 
#!/bin/sh

echo "hello world"
EOF
```

Create Dockerfile

```bash
cat <<-'EOF'> Dockerfile
FROM busybox
COPY /script.sh /
CMD ["/script.sh"]
EOF
```

Now trying to build an image from these;

example1 fails, because `script.sh` is not executable:

```bash
$ docker build --no-cache -t example1 .
$ docker run --rm example1
docker: Error response from daemon: OCI runtime create failed: container_linux.go:348: starting container process caused "exec: \"/script.sh\": permission denied": unknown.
``

example2: change the script to be executable, and build a new image. Now the image works:

```bash
$ chmod +x script.sh
$ docker build --no-cache -t example2 .
$ docker run --rm example2
hello world
```

example3; gives a "not found" error, even though the file exists, and is executable.
This is because it has the wrong line-endinges (DOS/Windows), but Linux gives a cryptic error for that situation;

```bash
$ unix2dos script.sh
$ docker build --no-cache -t example3 .
$ docker run --rm example3
standard_init_linux.go:190: exec user process caused "no such file or directory"
```


Closing this issue, but hope the above helps

**Comment by hurdissej** [comment]
You just made a new issue, fixed it, then closed the issue. Ingenius!

**Comment by fiantsev** [comment]
guys, this error may be related to corrupted images (in my case there was shutdown of a build machine during build process). Subsequent attempt to build lead to this problem. So i fixed this by removing all images (including hidden ones) because i cant rely of them anymore. My save-day command was:
```
docker rmi -f $(docker images -aq)
```
**!!! be careful - it will remove all your images !!!**

**Comment by dngrhm** [comment]
@fiantsev Thanks!!  This worked for me, but, instead of removing all the images, I used `docker images` to see the images built today and removed them using `docker rmi -f <list_of_image_ids>`

**Comment by oduvan** [comment]
@dnephin removing recent images helped as well. Thanks

**Comment by jeremiahsmall** [comment]
The "bent image" clue helped me. I got here from googling the error. Docker had crashed while it was upping a project, and the image it was on apparently got bent when it hung, or when I forced Docker to restart.

Removing the specific image that was corrupted with `docker rmi -f <myImageId>` was a good start, but the full resolution was to prune after that so it stopped bringing the image back from my local cache. `docker system prune`

**Comment by robjackstewart** [comment]
> @fiantsev Thanks!! This worked for me, but, instead of removing all the images, I used `docker images` to see the images built today and removed them using `docker rmi -f <list_of_image_ids>`

You sir, are fantastic!

**Comment by oguzhancelikarslan** [comment]
instead of writing a few commands, I reset to factory defaults through the docker desktop dashboard. 
This will clear everything so you can try what you want from scratch cleanly.

I strongly warn you about being careful - it will remove all of your images, containers, settings so on so forth!!!

**Comment by gnboorse** [comment]
In case anyone is still running into this issue (as I was)...
Apparently sometimes this output can appear when you use single quotes instead of double quotes in an `ENTRYPOINT` or `CMD` directive.

For instance, the exact same Dockerfile with `ENTRYPOINT ['/path/to/bin']` will produce an error message, but the one with 
`ENTRYPOINT ["/path/to/bin"]` will behave as expected, with no issues. For reference, this is the error message I was getting: 

```
docker: Error response from daemon: OCI runtime create failed: container_linux.go:349: starting container process caused "exec: \"/bin/sh\": stat /bin/sh: no such file or directory": unknown.
ERRO[0000] error waiting for container: context canceled 
```

Also, if you run `docker inspect <your image>`, the issue is pretty annoying to spot, as it's not immediately obvious that the entrypoint command is wrapped in single quotes. See this output from the docker inspect command:

```json
"Entrypoint": [
                "/bin/sh",
                "-c",
                "['/go/bin/app']"
            ],

```

Notice the single quotes.

I eventually stumbled on [this article](https://medium.com/engineering-zemoso/docker-single-quote-vs-double-quote-the-debug-hell-5149cf5177cb) which helped me tremendously, but only after I had debugged this issue for 3 and a half hours.

It would be really nice for this to be fixed somehow, but I'm not sure where to log a fresh issue.

**Comment by thaJeztah** [comment]
> Apparently sometimes this output can appear when you use single quotes instead of double quotes in an ENTRYPOINT or CMD directive.

Both the `ENTRYPOINT` and `CMD` Dockerfile instructions accept two forms; the _shell form_ and _exec form_. The _exec form_ expects a valid JSON array to be provided, and if the line is not valid JSON, it falls back to using the shell form.

In JSON, single quotes are _invalid_ (JSON only accepts double quotes), so in that case, your entrypoint is assumed to be in the shell format, and it will try to execute `[`, which is an executable; 

```bash
which [
/usr/bin/[
```

**Comment by gnboorse** [comment]
@thaJeztah  I understand why the issue occurs, just as you explained it. I just find the error messaging to be confusing in this case. There’s got to be a better way to communicate to users the unexpected side effects of this syntax.

**Comment by zhangguanzhang** [comment]
```
$docker inspect gcr.io/distroless/static-debian11:debug
            "Cmd": null,
            "Entrypoint": [
                "/busybox/sh"
            ]
```
change 
```
RUN mkdir -p test ...
```
to
```
RUN ["sh", "-c", "mkdir -p /app \
    && ln -sf /usr/local/bin/etcd* /app/etcd"]
```
works for me

**Comment by akx** [comment]
For me, this was actually caused by `/bin` (that had been a symlink to `/usr/bin`) having been inadvertently replaced by an actual empty `/bin` directory by an `ADD /some.tar.gz /` directive. That's something to take care of too.
